{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re\n",
    "import os\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def take_tail_fromcigar(strand, cig, seq):\n",
    "    '''\n",
    "    Function returns tail seqence based on the CIGAR and sequence of read\n",
    "    Input:\n",
    "        cig: CIGAR from sam file after alignment qith soft-clipping  [str]\n",
    "        seq: sequence of read started from 3' end, having posibble tail   [str]\n",
    "    Output:\n",
    "        sequence of the tail extracted from read sequence based on CIGAR [str]\n",
    "    Assumptions:\n",
    "        1. For tailed reads, CIGAR starts from \\d+S (softcliping code). It's about the number\n",
    "        of softclipped nucleotides. We assume that this is also the length of the polyA tail.\n",
    "        2. Sequence seq starts from 3' tail of read.\n",
    "    '''\n",
    "\n",
    "    cigar2 = str(cig)\n",
    "    if not \"S\" in cigar2:\n",
    "        return \"\"\n",
    "    else: \n",
    "        \n",
    "        \n",
    "        if strand =='-':      \n",
    "            search_result_minus = re.findall(\"[0-9]+S$\", cigar2)\n",
    "            if search_result_minus: \n",
    "                number_S = int(search_result_minus[-1].split('S')[0])\n",
    "               \n",
    "                tail = seq[-number_S:]\n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        elif strand == '+':\n",
    "            search_result_plus = re.findall(\"^[0-9]+S\", cigar2)\n",
    "            if search_result_plus:\n",
    "                number_S = int(cigar2.split('S')[0])\n",
    "                tail = seq[0:number_S] \n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        else:\n",
    "             return \"\"\n",
    "        \n",
    "\n",
    "##### Test the function take_tail_fromcigar() #####\n",
    "\n",
    "assert take_tail_fromcigar('-', '8S', 'UUUHSAAAAAAAAAA') =='AAAAAAAA'\n",
    "assert take_tail_fromcigar('-', '8S444M1S', 'AAAAAAAAAAUUUHS') =='S'\n",
    "assert take_tail_fromcigar('+', '8S444M3S', 'AAAAAAAAAAUUUHS') =='AAAAAAAA'\n",
    "assert take_tail_fromcigar('-', '67M7S', 'AAAAAAAAAAUUUHS') =='AAUUUHS'\n",
    "assert take_tail_fromcigar('-', '*', 'AAAAAAAAAAUUUHS') ==''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def take_tail_fromcigar_revstrand(strand, cig, seq):\n",
    "    '''\n",
    "    Function returns tail seqence based on the CIGAR and sequence of read\n",
    "    Input:\n",
    "        cig: CIGAR from sam file after alignment qith soft-clipping  [str]\n",
    "        seq: sequence of read started from 3' end, having posibble tail   [str]\n",
    "    Output:\n",
    "        sequence of the tail extracted from read sequence based on CIGAR [str]\n",
    "    Assumptions:\n",
    "        1. For tailed reads, CIGAR starts from \\d+S (softcliping code). It's about the number\n",
    "        of softclipped nucleotides. We assume that this is also the length of the polyA tail.\n",
    "        2. Sequence seq starts from 3' tail of read.\n",
    "    '''\n",
    "\n",
    "    cigar2 = str(cig)\n",
    "    if not \"S\" in cigar2:\n",
    "        return \"\"\n",
    "    else: \n",
    "        \n",
    "        \n",
    "        if strand =='+':      \n",
    "            search_result_minus = re.findall(\"[0-9]+S$\", cigar2)\n",
    "            if search_result_minus: \n",
    "                number_S = int(search_result_minus[-1].split('S')[0])\n",
    "               \n",
    "                tail = seq[-number_S:]\n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        elif strand == '-':\n",
    "            search_result_plus = re.findall(\"^[0-9]+S\", cigar2)\n",
    "            if search_result_plus:\n",
    "                number_S = int(cigar2.split('S')[0])\n",
    "                tail = seq[0:number_S] \n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        else:\n",
    "             return \"\"\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern_polyAU_forward = re.compile(\"^A{1,}T{1,}\") # reverse transcribed\n",
    "pattern_oligoU_forward = re.compile(\"^A{1,}\")\n",
    "pattern_polyA_forward = re.compile(\"^T{1,}$\")\n",
    "\n",
    "pattern_polyAU_reverse = re.compile(\"A{1,}T{1,}$\") # reverse transcribed\n",
    "pattern_oligoU_reverse = re.compile(\"T{1,}$\")\n",
    "pattern_polyA_reverse = re.compile(\"A{1,}$\")  \n",
    "\n",
    "def test_tail(strand, tail): ## poprawiona wersja january 2023\n",
    "    \"\"\" Tests for the type of tail. Patterns are prepared for analysis of R2 read, reverse transcribed:\n",
    "        base sequence start from the end of polyadenylation tail at 3' end of RNA,\n",
    "        where T means adenine, and A meand uridine in the 3' tail\"\"\"\n",
    "    if len(tail) > 0:\n",
    "        if (\"G\" in tail or \"C\" in tail):\n",
    "            return \"mix_GC\"\n",
    "        else:\n",
    "            \n",
    "            if strand == '-' or strand == 'unmapped': ## podmianka +*-\n",
    "\n",
    "                if \"A\" in tail:\n",
    "                    if pattern_polyAU_forward.fullmatch(tail): \n",
    "                        return \"polyAU\"\n",
    "                    elif pattern_oligoU_forward.fullmatch(tail): \n",
    "                        return \"oligoU\"\n",
    "                    else:\n",
    "                        return \"mixed_tail\"\n",
    "                elif pattern_polyA_forward.fullmatch(tail):\n",
    "                        return \"polyA\"\n",
    "                else: \n",
    "                    return \"no_asigned_tail\"\n",
    "            elif strand == '+': ## podmianka +*-\n",
    "                if \"T\" in tail:\n",
    "                    if pattern_polyAU_reverse.fullmatch(tail): \n",
    "                        return \"polyAU\"\n",
    "                    elif pattern_oligoU_reverse.fullmatch(tail): \n",
    "                        return \"oligoU\"\n",
    "                    else:\n",
    "                        return \"mixed_tail\"\n",
    "                elif pattern_polyA_reverse.fullmatch(tail):\n",
    "                        return \"polyA\"\n",
    "                else: \n",
    "                    return \"no_asigned_tail\"\n",
    "            else:\n",
    "                return \"else_tail\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern_mRNA = re.compile(\"spac|spap|spbp|spcc|spcp|spbc\")\n",
    "pattern_tRNA = re.compile(\"trna\")\n",
    "pattern_mito = re.compile(\"spmit\")\n",
    "pattern_rRNA = re.compile(\"rrna\")\n",
    "pattern_snRNA = re.compile(\"snrna\")\n",
    "pattern_snoRNA = re.compile(\"snorna\")\n",
    "pattern_ncRNA = re.compile(\"ncrna\")\n",
    "\n",
    "\n",
    "def test_RNA_type(feature_name):\n",
    "    \"\"\"tests for the type of tail \"\"\"\n",
    "    if pattern_mRNA.search(feature_name): # for example AATT\n",
    "        return \"mRNA\"\n",
    "    elif pattern_tRNA.search(feature_name): # for example AATT\n",
    "        return \"tRNA\"\n",
    "    elif pattern_mito.search(feature_name): # for example AATT\n",
    "        return \"mito_RNA\"\n",
    "    elif pattern_rRNA.search(feature_name): # for example AATT\n",
    "        return \"rRNA\"\n",
    "    elif pattern_snRNA.search(feature_name): # for example AATT\n",
    "        return \"snRNA\"\n",
    "    elif pattern_snoRNA.search(feature_name): # for example AATT\n",
    "        return \"snoRNA\"\n",
    "    elif pattern_ncRNA.search(feature_name): # for example AATT\n",
    "        return \"ncRNA\"\n",
    "    else: #only A's\n",
    "        return \"other_type\"\n",
    "\n",
    "assert test_RNA_type(\"spncrna\") == \"ncRNA\"\n",
    "assert test_RNA_type(\"sspcc\") == \"mRNA\"\n",
    "assert test_RNA_type(\"AAAAA\") == \"other_type\"\n",
    "assert test_RNA_type(\"TAATTT\") == \"other_type\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_tails_from_CIGAR_new(bam_file_path, references_bed_path,\n",
    "                            output_path='output_tailsfromcigar.csv'):\n",
    "    #sam_file = pd.read_csv(sam_file_path, index=False)\n",
    "    #sam_file.to_csv('sam_input.sam',index=False)\n",
    "   # references_file = pd.read_csv(references_bed_path)\n",
    "    #references_file.to_csv('references.bed', sep ='\\t', indexl=False)\n",
    "    \n",
    "\n",
    "\n",
    "    os.system('samtools view -F 256 {} > primaalign.sam'.format(bam_file_path))\n",
    "    #os.system('cp {}_primaalign.sam temporary.sam'.format(bam_file_path))\n",
    "\n",
    "    os.system(\"awk -F'\\t' '$6~/S/'  primaalign.sam > temporary_awk.sam\")\n",
    "\n",
    "    os.system(\"sam2bed temporary_awk.sam temporary_awk.bed\") \n",
    "\n",
    "    os.system(\"bedtools sort -i temporary_awk.bed > temporary_awksorted.bed\") \n",
    "\n",
    "\n",
    "    os.system(\"bedtools closest -s -d -t first -a temporary_awksorted.bed -b {} >  closest.bed\".format(references_bed_path))  \n",
    "\n",
    "    os.system(\"awk '{print  $1, $2, $3, $4, $5, $6,  $8, $9, $10, $17  }' closest.bed > closest_lesscol.bed\")\n",
    "\n",
    "    os.system(\"awk '{ print $1,$6, $10 }' temporary_awk.sam > primaalign_awk_col6_10.csv\")\n",
    "   \n",
    "    df_closest = pd.read_csv('closest_lesscol.bed', sep = ' ',\n",
    "                             names = ['chr',  'start', 'stop', 'read_id', 'num0', 'strand', 'start_gene',\n",
    "                                      'stop_gene', 'gene_id',   'distance_clos'])\n",
    "    df_minisam =  pd.read_csv('primaalign_awk_col6_10.csv', sep = ' ', \n",
    "                     names = [ 'read_id', 'cigar', 'seq'])\n",
    "    cigar = pd.merge(df_closest , df_minisam, on = 'read_id')\n",
    "    cigar['gene'] = cigar['gene_id'].apply(lambda x: x.split(':')[0])\n",
    "    \n",
    "    cigar['tail_fromcigar'] = cigar.apply(lambda kol: take_tail_fromcigar_revstrand(kol.strand, \n",
    "                                                                          kol.cigar,kol.seq), axis = 1)\n",
    "    cigar['tail_type'] = cigar.apply(lambda kol: test_tail(kol.strand, kol.tail_fromcigar), axis = 1)\n",
    "    cigar['tail_len'] = cigar['tail_fromcigar'].apply(lambda x: len(str(x)))\n",
    "    cigar['id_lower'] = cigar['gene_id'].apply(lambda x: x.lower())\n",
    "    cigar['RNA_type'] = cigar['id_lower'].apply(lambda x: test_RNA_type(x))\n",
    "    cigar.to_csv(output_path, index = False)\n",
    "   # return cigar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TEST the code\n",
    "detect_tails_from_CIGAR_new(bam_file_path = '../mapped/WT_2.bam', \n",
    "                            references_bed_path = '../reference/newSP_gene.bed',\n",
    "                            output_path='output_tailsfromcigar.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
